### PR TITLE
Merge the omrargscan library from OpenJ9

### PR DESCRIPTION
**Pull request by mthom**
Merge omrargscan.[c|h] in support of porting the refactored OpenJ9 shared cache into OMR.

Issue: #4610

Signed-off-by: Mark Thom <mark.thom@unb.ca>

### CODE REVIEW

File: port/CMakeLists.txt [hunk 1]
```diff
@@ -163,7 +163,8 @@ if(OMR_ENV_DATA64)
 endif()
 
 list(APPEND OBJECTS
-	omrheap.c
+        omrargscan.c
```
Comment (hangshao0) [inline port/CMakeLists.txt]: incorrect indent ?

File: port/common/omrargscan.c [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+
+/* Returns the trimmed input string, removing leading and trailing whitespace.
+ * Returned string is malloc-ed and must be freed.
+ * @param portLibrary The portLibrary to use
+ * @param input The buffer to trim.
+ */
+char * trim(OMRPortLibrary *portLibrary, char * input)
+{
+	BOOLEAN whitespace = TRUE;
+	uintptr_t index;
```
Comment (hangshao0) [inline port/common/omrargscan.c]: `index` is not initialized.

File: port/common/omrargscan.c [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+/* Answer string up to specified delimiter or end of line */
+/* Return value is malloc-ed and must be freed by the sender */
+char *scan_to_delim(OMRPortLibrary *portLibrary, char **scan_start, char delimiter)
+{
+	char *scan_string = *scan_start, *subString;
+	uintptr_t i = 0;
+
+	while (scan_string[i] && (scan_string[i] != delimiter))
+		i++;
```
Comment (hangshao0) [inline port/common/omrargscan.c]: Please put `i++` inside `{` and `}`

File: port/common/omrargscan.c [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+	}
+
+	return buf;
+}
+
+/* Answer string up to specified delimiter or end of line */
+/* Return value is malloc-ed and must be freed by the sender */
+char *scan_to_delim(OMRPortLibrary *portLibrary, char **scan_start, char delimiter)
+{
+	char *scan_string = *scan_start, *subString;
```
Comment (hangshao0) [inline port/common/omrargscan.c]: Please define `*subString` in a new line.

File: port/common/omrargscan.c [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+	}
+	return subString;
+}
+
+/*
+ * Print an error message indicating that an option was not recognized
+ */
+void scan_failed(OMRPortLibrary * portLibrary, const char* module, const char *scan_start)
+{
+  portLibrary->tty_printf(portLibrary,
```
Comment (hangshao0) [inline port/common/omrargscan.c]: incorrect indent ?

File: port/common/omrargscan.c [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+
+		default:
+			whitespace = TRUE;
+		}
+	}
+
+	return buf;
+}
+
+/* Answer string up to specified delimiter or end of line */
```
Comment (hangshao0) [inline port/common/omrargscan.c]: Please add doxygen comment describing what the function does, what are the input parameters and return value. Similarly for other functions in this file.

**Comment by youngar** [comment]
@mthom I think it would be better if this functionality was added to omrutil library.

**Comment by rwy0717** [comment]
The functions ought to be prefixed with `omrargscan_` or similar.

My main concern is, why are we adding more option parsing to OMR? Is there another PR that's using this code?

**Comment by mthom** [comment]
> 
> 
> The functions ought to be prefixed with `omrargscan_` or similar.
> 
> My main concern is, why are we adding more option parsing to OMR? Is there another PR that's using this code?

Yes, the shared cache port uses this code.

**Comment by rwy0717** [comment]
Do you think you could point me at the code using this new library?

**Comment by DmonteDD** [comment]
> Do you think you could point me at the code using this new library?

The functionality introduced in this PR is not referred by any other/later Shared Cache PR's. This functionality provides convenience for future consumer's of Shared Cache. At present, we have exercised them in our prototype implementation.

**Comment by DmonteDD** [comment]
> @mthom I think it would be better if this functionality was added to omrutil library.

I have ported into omrutil. Thanks :)

**Comment by DmonteDD** [comment]
> The functions ought to be prefixed with `omrargscan_` or similar.
> 
> My main concern is, why are we adding more option parsing to OMR? Is there another PR that's using this code?

I have prefixed all functions with `omr_` . Thanks

**Comment by youngar** [comment]
Thanks for fixing up this PR @DmonteDD!
@genie-omr build all

**Comment by youngar** [comment]
@genie-omr build win

**Pull request merged**